[Batch mode] Write supplementary output file map. rdar://problem/38157859

davidungar opened the pull request:
<!-- What's in this pull request? -->
When there are lots of supplementary outputs (only happens in batch mode, but the code is simpler if it does not care), use a filemap to communicate them from driver to frontend.

There are many changes here:

- OutputFileMap.* and Types.h move from the driver into the frontEnd so that frontEnd can reuse the file map format and parsing logic.
- `OutputFileMap` parsing reports more information about errors.
- Some behavior about using various filelists and adding corresponding arguments moves to `JobContext`, where all the information already resides.
- The driver gains a new kind of `FileListInfo`, `SupplementaryOutput`.
- The existing driver code to repartition a batch if the command line is too big is kept as a recovery mechanism in case the heuristic used to decide whether to employ a supplementary output file map reports a false negative. In that case, a message is output to `stderr` in debug mode.
- The invocation creation code for compile jobs is unified across single-, batch-, and standard-compiles.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Commit:
Write out filemaps for supplementary outputs.

graydon commented on include/swift/Driver/Job.h:
```diff
@@ -206,6 +206,9 @@ class CommandOutput {
   StringRef getAnyOutputForType(types::ID type) const;
 
+  /// Return the whole derived output map.
+  const OutputFileMap &getDerivedOutputMap() const;
```
The derived OFM is shared across the whole compilation; it can be written once per compilation (not once per job). There's an accessor for it on the Compilation object.

davidungar replied:
Thanks a lot for doing this so quickly. I'll change tacks right away.

davidungar replied:
After looking at one OFM file for all jobs, I concluded that it was worth exploring separate OFMs for two reasons: 
1. The code in the driver seems a bit cleaner because it need not add state to the Compilation to remember if the file map has been written yet (it's lazy) and what the path name was, and 
2. The cost for each batch mode frontend to read a global filemap becomes quadratic: the # of batches * the # of primaries in the whole compilation.

So, I've gone this way for now.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test

davidungar commented:
Because I had to move OutputFileMap.h and Types.h a number of files where only changed in their include statements. But git-clang-format changed other lines in them, too. (Apparently, they weren't already processed by git-clang-format.) These changes put me in a quandary: on one hand, I could undo them, which would make this review easier. But on the other hand, any subsequent change that reformatted the files would be harder. (Or maybe there's a better way to run git-clang-format that I don't know.) I have taken the former course, but can go back and undo all spurious formatting changes if you like. I don't believe there any files which were completely unchanged, yet reformatted.

davidungar commented:
There were some routines in the driver that grew large enough with the addition this functionality that I felt it worthwhile to break them up. In addition, many of the pieces seemed to me to belong in the `JobContext` class, because they kept accessing the `context` instance variable of the `ToolChain`.
Embedding behavior in this class represents a shift from the prior state of the code, but sometimes I feel that the code shouts at me that what I thought a concept represented was not quite what it needs to represent to get the job done. See what you think.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@jrose-apple & @graydon Although I haven't smoked this change yet, I think it's in good enough shape to be reviewed at this point.

jrose-apple commented on lib/Driver/Compilation.cpp:
```diff
@@ -36,9 +37,9 @@
-#include "llvm/Support/raw_ostream.h"
 #include "llvm/Support/Timer.h"
 #include "llvm/Support/YAMLParser.h"
+#include "llvm/Support/raw_ostream.h"
```
clang-format does a case-sensitive ordering? :-( Okay…

davidungar replied:
Whew! :)

jrose-apple commented on lib/Driver/Compilation.cpp:
```diff
@@ -804,6 +809,9 @@ namespace driver {
           // aggressively by doubling partition count / halving size.
           NumPartitions *= 2;
+#ifndef NDEBUG
+          llvm::errs() << "Should have used a supplementary output file map.\n";
```
I don't think we should be writing random things to `llvm::errs()` without a specific flag. `llvm::dbgs()` might be a good alternative (requires a flag to turn *on).*

davidungar replied:
I was wondering about that. Changed to dbgs. Keep the NDEBUG conditional?

jrose-apple replied:
Ah, I think using the LLVM `DEBUG` macro is sufficient, no need for `NDEBUG`.

davidungar replied:
Cool! Adding a DEBUG_TYPE called BATCH_MODE, hope that's right.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -17,25 +17,25 @@
 #include "swift/Basic/Statistic.h"
+#include "swift/Basic/TaskQueue.h"
+#include "swift/Basic/Version.h"
+#include "swift/Config.h"
```
Nitpick: this isn't right either. The files in `swift/` and not in a subdirectory all go together. (Whether that "together" is the beginning or the end I'm not sure, but I'd be inclined to say the end.)

davidungar replied:
I'm confused. Am I running git-clang-format wrong?
BTW, here is what it was on 3/2 before I touched it:
#include "ToolChains.h"
#include "swift/Strings.h"
#include "swift/AST/DiagnosticEngine.h"
#include "swift/AST/DiagnosticsDriver.h"
#include "swift/AST/DiagnosticsFrontend.h"
#include "swift/Basic/LLVM.h"
#include "swift/Basic/TaskQueue.h"
#include "swift/Basic/Version.h"
#include "swift/Basic/Range.h"
#include "swift/Basic/Statistic.h"
#include "swift/Driver/Action.h"
#include "swift/Driver/Compilation.h"
#include "swift/Driver/Job.h"
#include "swift/Driver/OutputFileMap.h"
#include "swift/Driver/PrettyStackTrace.h"
#include "swift/Driver/ToolChain.h"
#include "swift/Option/Options.h"
#include "swift/Option/SanitizerOptions.h"
#include "swift/Parse/Lexer.h"
#include "swift/Config.h"
#include "llvm/ADT/APInt.h"
#include "llvm/ADT/DenseSet.h"
#include "llvm/ADT/STLExtras.h"
#include "llvm/ADT/SmallString.h"
#include "llvm/ADT/StringSwitch.h"
#include "llvm/Config/config.h"
#include "llvm/Option/Arg.h"
#include "llvm/Option/ArgList.h"
#include "llvm/Option/OptTable.h"
#include "llvm/Support/Debug.h"
#include "llvm/Support/ErrorHandling.h"
#include "llvm/Support/FileSystem.h"
#include "llvm/Support/Host.h"
#include "llvm/Support/MD5.h"
#include "llvm/Support/Path.h"
#include "llvm/Support/PrettyStackTrace.h"
#include "llvm/Support/raw_ostream.h"

jrose-apple replied:
Oh, right, I think we've had this discussion before. https://bugs.llvm.org/show_bug.cgi?id=34679

…carry on.

(And for good measure, https://bugs.llvm.org/show_bug.cgi?id=34678)

jrose-apple commented on lib/Driver/ToolChain.cpp:
```diff
@@ -47,6 +48,47 @@ const char *ToolChain::JobContext::getAllSourcesPath() const {
+
+bool ToolChain::JobContext::shouldUsePrimaryInputFileList() const {
+  // SingleCompile's must not return true because then all inputs erroniously
+  // end up in primary flielist.
```
Typos: "erroniously" and "flielist".

davidungar replied:
Unintentional meta-joke.

davidungar replied:
Tx

jrose-apple commented on lib/Driver/ToolChain.cpp:
```diff
@@ -47,6 +48,47 @@ const char *ToolChain::JobContext::getAllSourcesPath() const {
+bool ToolChain::JobContext::shouldUseSupplementaryOutputFileList() const {
+  if (Args.hasArg(options::OPT_driver_use_filelists))
+    return true;
+  static const uint UpperBoundOnSupplementaryOutputFileTypes =
```
What is `uint`?

davidungar replied:
Changed to unsigned.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -104,11 +101,10 @@ addOutputsOfType(ArgStringList &Arguments,
+static void
+addCommonFrontendArgs(const ToolChain &TC, const OutputInfo &OI,
+                      const CommandOutput &output, const ArgList &inputArgs,
+                      ArgStringList &arguments) {
```
Nitpick: Please don't reformat things if you're not actually changing them.

davidungar replied:
How do I run git-clang-format so it only reformats the lines I changed?

davidungar replied:
Fixed.

jrose-apple replied:
I thought git-clang-format did that automatically. (That's what it says in its help output.) Is that not the case?

davidungar replied:
I'll be watching it carefully!

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +353,186 @@ ToolChain::constructInvocation(const CompileJobAction &job,
   return II;
 }
 
+const char *ToolChain::JobContext::computeFrontendModeForCompile() const {
```
This one still makes me uncomfortable. Why does JobContext know about fontend flags? The same struct is used for all the different kinds of jobs, but this one's only relevant for compile jobs.

davidungar replied:
Interesting. This function does have problems: it's doing two things, filtering out Immediate and REPL compiles, then mapping primary output type to compilation mode arguments. I agree that it's misplaced. Where does the knowledge of modes belong? How about in CompileJobAction? Would that be better?

jrose-apple replied:
(planning to talk about this in person today)

jrose-apple replied:
(We decided that it's okay to leave these on JobContext as long as they have "Frontend" in the name.)

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +353,186 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+  switch (Output.getPrimaryOutputType()) {
+  case types::TY_Object:
+    return "-c";
+    break;
```
No need for `break` on any of these.

davidungar replied:
Sorry, moving too fast.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +353,186 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+  }
+}
+
+void ToolChain::JobContext::addCompileArgumentsEitherOnCommandLineOrInFilelists(
```
This definitely does not belong on JobContext. It is not JobContext's job (heh) to populate argument lists, and (same as above) I don't think it makes sense for JobContext to know about frontend options at all.

davidungar replied:
Let's talk about this offline. We can hash out the utility of defining JobContext's job according to various criteria. There's something I don't yet understand going on here.

davidungar replied:
Move to CompileJobAction, or where??

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -214,19 +210,26 @@ static void addCommonFrontendArgs(const ToolChain &TC,
     arguments.push_back("-color-diagnostics");
+}
 
+static void addModuleDocPathArgIfNecessary(const CommandOutput &output,
```
Can you explain why this got factored out? It was already in a "common" location and now it gets invoked multiple times.

davidungar replied:
The problem is that it should not run for a compile job with a supplementary output file map. The options that I can think of are to:
1. add a check where it was to avoid outputting it in that case,
2. pass in an extra parameter to the function where it used to be, or
3. yank it out of the add-common-arguments function and call it separately.

I chose #3, but I'm happy to revisit it. I'm not thrilled with any of the possibilities.

jrose-apple replied:
How does this happen for all the other supplementary outputs? Or do we just…not have any besides modules and serialized diagnostics?

davidungar replied:
I think these are the only two that were in the common routine. The others were inserted from compile-job-specific code. They were done via control flow, IIRC, so that's why I did it that way for these two.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -451,3 +454,77 @@ void SupplementaryOutputPathsComputer::deriveModulePathParameters(
+  SupplementaryOutputPaths paths;
+  if (!map)
+    return paths;
+  static const std::vector<std::pair<types::ID, std::string &>> typesAndStrings{
```
Please use a plain old C array for this rather than a std::vector, since it won't change sides. (It's also *definitely* not `static`, since you include references to the local variable!)

jrose-apple replied:
If you really want to make it static, you can use pointers-to-members here.

davidungar replied:
Thanks for the catch on the *static* -- d'oh! Not sure what 'change sides' means, but I'm happy to make it a C array, with some sentinel. Don't understand why a C array is better.

jrose-apple replied:
Oops, that was supposed to be "size". And you don't need a sentinel; C arrays know their sizes. That means the whole thing can be a single stack allocation.

davidungar replied:
Thanks. Fixed!

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -451,3 +454,77 @@ void SupplementaryOutputPathsComputer::deriveModulePathParameters(
+                               options::OPT_emit_tbd_path)) {
+    // Since the table is only used for a diagnostic, spend the time to recreate
+    // it and save the complexity of passing it in.
+    std::unique_ptr<llvm::opt::OptTable> opts(createSwiftOptTable());
```
Please don't. It's okay to use string constants for command-line arguments in diagnostics, especially if it's a diagnostic that will only show up for people writing tests.

davidungar replied:
OK, but I don't understand why. Is there some performance criticality here I don't know about?

jrose-apple replied:
In this case it's more code complexity. This is jumping through hoops to get something that's ultimately a static string, and on top of that it's a string that's unlikely to ever change.

davidungar replied:
OK, changed!

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -451,3 +454,77 @@ void SupplementaryOutputPathsComputer::deriveModulePathParameters(
+  OutputFileMap OFM;
+  std::string error = OFM.loadFromBuffer(std::move(buffer.get()), "");
+  if (!error.empty()) {
+    // TODO: emit diagnostic with error string
```
Is this TODO still relevant?

davidungar replied:
Thanks for catching this.

jrose-apple commented on lib/Frontend/OutputFileMap.cpp:
```diff
@@ -108,21 +100,42 @@ void OutputFileMap::dump(llvm::raw_ostream &os, bool Sort) const {
+      continue;
+    for (auto &typeAndOutputPath : *outputMap) {
+      printQuotedString(input);
+      os << " :\n";
```
Nitpick: typical YAML style would not put a space here.

davidungar replied:
Tx. Fixed.

jrose-apple commented on lib/Frontend/OutputFileMap.cpp:
```diff
@@ -108,21 +100,42 @@ void OutputFileMap::dump(llvm::raw_ostream &os, bool Sort) const {
+void OutputFileMap::write(llvm::raw_ostream &os,
+                          ArrayRef<StringRef> inputs) const {
+  auto printQuotedString = [&os](StringRef s) -> void {
+    os << "\"" << s << "\"";
```
Please use raw_ostream's `write_escaped` method for the string. (I get away with not doing that in some other places because the strings are all identifiers, but paths are not that nice.)

davidungar replied:
Cool! Thanks.

jrose-apple commented on lib/Frontend/Types.cpp:
```diff
@@ -27,41 +26,34 @@ struct TypeInfo {
-  { NAME, FLAGS, TEMP_SUFFIX },
-#include "swift/Driver/Types.def"
+#define TYPE(NAME, ID, TEMP_SUFFIX, FLAGS) {NAME, FLAGS, TEMP_SUFFIX},
+#include "swift/Frontend/Types.def"
```
Another file of "please don't reformat things you aren't changing".

davidungar replied:
But I *did* change this file: I removed the driver namespace, because it moved to the frontend.
I thought the rule was: "Don't reformat files you don't change, but it's OK to reformat a file you do change." As I tried to explain in the PR preamble: there was a dilemma. I could undo the git-clang-format changes to ease reviewing, but then the next person who changed the file and ran git-clang-format would just change it again. What am I missing here?

jrose-apple replied:
My understanding was that git-clang-format is only supposed to edit the lines you changed, but when you move a file it thinks the whole thing has changed. That's why it won't be a problem for the next person. But (cf. previous comment) maybe I'm wrong about that.

I suppose you could trick it into not doing this behavior by moving the files first in a separate commit that breaks things, then making the changes, and then squashing the rest of it back.

davidungar replied:
I've gone back and manually minimized the diffs.

jrose-apple commented on test/Driver/batch_mode_with_supplementary_filelist.swift:
```diff
@@ -0,0 +1,15 @@
+//
+// RUN: %swiftc_driver -enable-batch-mode -driver-use-filelists -j2 %t/main.swift  %t/file-01.swift %t/file-02.swift %t/file-03.swift  -o %t/a.out
+// RUN: %t/a.out | %FileCheck %s -check-prefix=CHECK-HELLO-WORLD
+// CHECK-HELLO-WORLD: Hello, World!
```
Please split the executable part of this test into a separate file, so we can mark it with `// REQUIRES: executable_test`. You'll also have to use `%target-build-swift` instead of `%swiftc_driver` to get the right build settings, and `%target-run` to make sure the resulting executable is run correctly for non-host platforms like iOS.

davidungar replied:
Thanks. I'll try to study Testing.md.

davidungar replied:
Did split up the test as suggested.

jrose-apple reviewed (commented):
Well, it looks like we've hit another of the many-comment PRs :-), and we'll probably need to work out the JobContext thing in person. Maybe not today, though.

davidungar commented:
@swift-ci please smoke test

Commit:
Add -driver-force-one-batch-repartition and enhance  batch_mode-overlong_argv test.

davidungar commented:
@swift-ci please smoke test

Commit:
WIP

davidungar commented on include/swift/Frontend/OutputFileMap.h:
```diff
@@ -38,24 +38,27 @@ class OutputFileMap {
-  loadFromPath(StringRef Path, StringRef workingDirectory);
+  ///
+  /// \return a non-empty error description on error.
+  std::string loadFromPath(StringRef Path, StringRef workingDirectory);
```
I agree! But how do I use ErrorOr to return an arbitrary string instead of an error code?

jrose-apple replied:
Ah, the *correct* thing to do is probably to define your own subclass of llvm::ErrorInfo (something that the Serialization library does)…but there's also an llvm::StringError. That'll look like `llvm::make_error<llvm::StringError>("bad input")`, I think.

davidungar replied:
Thanks! I'll look at StringError--did not know about it.

Commit:
OFM fixes

Commit:
Debug macro

Commit:
Formatting

davidungar commented:
@swift-ci  please smoke test

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test

Commit:
Fix test

Commit:
Implement -index-file-path

davidungar commented:
@swift-ci  please smoke test

davidungar commented:
@swift-ci please smoke test

Commit:
Fix test.

Commit:
Patched up Driver/filelists.swift test.

Commit:
Fix test Driver/merge-module.swift

Commit:
Don't create Swift option table for a diagnostic.

davidungar commented:
@swift-ci please smoke test

Commit:
Fix driver-compile.swift test.

jrose-apple commented on include/swift/Frontend/Types.h:
```diff
@@ -19,11 +19,10 @@
 
 namespace swift {
-namespace driver {
 namespace types {
```
Hm. `swift::types` makes me think of types in the language rather than file types. Maybe we should rename the namespace to `file_types`?

(yes, it's weird that [namespaces are the only thing in the project that uses lower_snake_case](https://github.com/apple/swift/blob/master/include/swift/Basic/StringExtras.h#L70))

davidungar replied:
Agree! Separate PR? Before or after this one??

jrose-apple replied:
Ah, I figured that since you were removing `driver` most of the use sites would have to be updated anyway, but that's probably not true. "Before" is probably the most correct thing to do, but either way is fine as long as we don't forget about it.

davidungar replied:
OK, tx. Will do it after--really want to land this one (as usual).

jrose-apple commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -30,6 +30,9 @@ def primary_filelist : Separate<["-"], "primary-filelist">,
   HelpText<"Specify primary inputs in a file rather than on the command line">;
 def output_filelist : Separate<["-"], "output-filelist">,
   HelpText<"Specify outputs in a file rather than on the command line">;
+def supplementary_output_filemap : Separate<["-"], "supplementary-output-filemap">,
```
Nitpick: The term is "output file map", while `-output-filelist` really is "filelist" (for the silly reason that the equivalent linker flag is spelled `-filelist` rather than `-file_list`, when the Apple linker has plenty of other underscore-separated flags). So please use `file-map` here.

Additionally, is it worth putting "supplementary" in the command-line flag? I know we haven't decided yet whether it makes sense to pull primary output files from here too, and it definitely makes sense to use "supplementary" in all the frontend function names. And it doesn't *really* matter because it's a frontend flag. But still, what do you think about just using `-output-file-map`?

davidungar replied:
Changed `filemap` to `file-map` or `file_map` as needed.

I like the idea of just using `-output-file-map`, but prefer to do it as part of a larger merge, unifying primary and supplementary outputs. Even if not, I would defer it to a next PR.

jrose-apple commented on lib/Driver/Compilation.cpp:
```diff
@@ -36,12 +37,14 @@
 
 #include "CompilationRecord.h"
 
+#define DEBUG_TYPE "BATCH_MODE"
```
Nitpick: it looks like we're not consistent about how we use this macro ("lower-kebab-case"? "UpperCamelCase"? "Title Case"? "Regular sentence case"?), but "UPPER_SNAKE_CASE" is not one of the ones we've used so far. "lower-kebab-case" looks like it makes the most sense, since it's going to be used on the command-line as `-Xllvm -debug-only=foo-bar`.

davidungar replied:
Thanks! fixed. Never heard of "kebab-case" before, but it's a wonderful term. If languages only allowed blanks in identifiers we wouldn't need three ways--but would have other problems. (REX (or was it REXX)?) not only allowed blanks in IDs, but supplied a value for an uninitialized variable: the name of the variable as a string! It was great for small scripts, batch-processed from card decks, where one forgot to quote a string. I never used it, though, and am not advocating such.

jrose-apple commented on include/swift/Driver/ToolChain.h:
```diff
@@ -204,6 +242,10 @@ class ToolChain {
                                          bool shared=true) const;
 
+  /// For an index data job,
+  /// only make the index-file-path ones primary.
```
Thinking about this more, I'm pretty sure this function is the wrong answer anyway. Whatever's doing indexing should just construct jobs appropriately so that the one input is marked as primary. But you don't have to change that in this PR.

davidungar replied:
There's this test, driver-index.swift:
`// RUN: %swiftc_driver -driver-print-jobs -index-file -index-file-path %s %S/Inputs/SwiftModuleA.swift %s -o %t.output_for_index -index-store-path %t.index_store -module-name driver_index 2>&1 | %FileCheck %s`.
It names %s *twice* in the command line, and expects the frontend to see one %s, as the first input, as a primary input.
I introduced this function to preserve this behavior. It would be great if we could change the test!

jrose-apple replied:
New round of comments where I've given up on improving `-index-file`: still, this is an implementation comment rather than an interface comment.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -84,7 +84,7 @@ void Driver::parseDriverKind(ArrayRef<const char *> Args) {
   // However, the driver kind may be overridden if the first argument is
   // --driver-mode.
-  if (!Args.empty()) {
+  if (Args.size() > 0) {
```
Why this change? `empty()` is idiomatic C++ like it is in Swift because `size()` may be expensive to compute.

davidungar replied:
Thanks for catching it. It was a mistake: I compared the current version to an older one and picked the wrong old version to compare it to.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -987,6 +992,17 @@ void Driver::buildInputs(const ToolChain &TC,
       }
 
+      // Apparently the convention is that the index file path is repeated.
+      // Don't try to add it in twice.
```
It's not repeated. The convention is that the index file path specifies one of the input files; it is not itself an input file. It's questionable whether this is the *right* convention, but it *must* be separate, not "may be separate".

davidungar replied:
What confuses me is the input ordering:
The driver-index.swift test does essentially this:
`swiftc -### -index-file -index-file-path .../driver-index.swift  .../SwiftModuleA.swift -module-name foo`
and the frontend sees:
`swift -frontend -typecheck -primary-file .../driver-index.swift .../SwiftModuleA.swift ... -module-name foo -o foo`

How does that `driver-index.swift` get to be the first input?

davidungar replied:
Checked with @akyrtzi and found out that the key requirement is to make the index file path primary. The reordering in the status quo is not required. See `Driver::buildInputs` and `ToolChain::canCompileInputArgumentBePrimary` for my current solution. Tx.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1102,7 +1118,8 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
     OI.CompilerMode = OutputInfo::Mode::StandardCompile;
-    if (Args.hasArg(options::OPT_whole_module_optimization))
+    if (Args.hasArg(options::OPT_whole_module_optimization,
+                    options::OPT_index_file))
```
This seems wrong. The whole point up till now is that index-file is going to work like single-file mode. Maybe just make jobs non-batchable if they're indexing?

davidungar replied:
I agree it seems wrong. When I tried it the other way, something else complained about the number of output files not matching the number of inputs.
I can look at it again. The other question in my mind was does an indexing job only index the primary file? Or does it index the whole module??

davidungar replied:
Please take a look at Driver::buildInputs in Driver.cpp. The issue is how best to downstream-signal to make the index file path a primary. What I have done is let the argument carry the information, by using the argument with that option in the list. Seems cleanest to me, but it could also be done by adding state somewhere else. Let me know how you feel.

jrose-apple replied:
I really think the right answer is to have -index-file force single-file mode even if the user passes -wmo, and then -index-file-path controls the one job that gets made.

jrose-apple replied:
(After talking to @benlangmuir I no longer think this is the case, though it makes me sad because other things are probably wrong because of it. But it's not something that has to change in this PR.)

jrose-apple commented on lib/Driver/Job.cpp:
```diff
@@ -240,6 +240,10 @@ StringRef CommandOutput::getAnyOutputForType(types::ID Type) const {
 }
 
+const OutputFileMap &CommandOutput::getDerivedOutputMap() const {
+  return DerivedOutputMap;
```
Hm, is the Job really the right way to expose this? It's a shared structure and will contain more information than just what's in the Job.

davidungar replied:
I don't really like that this whole function is file-static. Maybe if you can help me figure out where it really belongs this issue will be clearer. I suspect I did it that way because it represented the least amount of encapsulation-breaking needed, given where the client was.

davidungar replied:
What did we decide to do about this?

jrose-apple replied:
Maybe the whole "write the subset of the DOM based on this Job's inputs" should be pushed into Job (taking an `llvm::raw_ostream &`). What do you think?

jrose-apple commented on lib/Driver/ToolChain.cpp:
```diff
@@ -37,8 +38,8 @@ ToolChain::JobContext::JobContext(Compilation &C,
                                   const OutputInfo &OI)
-  : C(C), Inputs(Inputs), InputActions(InputActions), Output(Output),
-    OI(OI), Args(C.getArgs()) {}
+: C(C), Inputs(Inputs), InputActions(InputActions), Output(Output),
```
This looks like it got mis-formatted. Or if it's what clang-format does, clang-format is clearly wrong!

davidungar replied:
I must have manually set it back to correspond with a prior SHA: 26d6502.
(Hard to know when to leave git-clang-format alone vs when to avoid changes.)

jrose-apple commented on lib/Driver/ToolChain.cpp:
```diff
@@ -21,12 +21,13 @@
 #include "swift/Driver/Compilation.h"
 #include "swift/Driver/Driver.h"
 #include "swift/Driver/Job.h"
+#include "swift/Option/Options.h"
```
This doesn't seem to be used in your diff. Or were we implicitly including it before and now we're not?

davidungar replied:
Great catch! Fixed.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -37,8 +38,46 @@ using namespace swift;
+  if (Args.hasArg(options::OPT_driver_use_filelists))
+    return true;
+  static const unsigned UpperBoundOnSupplementaryOutputFileTypes =
+  types::TY_INVALID;
```
Nitpick: needs indentation…

davidungar replied:
Will let git-clang-reformat.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -37,8 +38,46 @@ using namespace swift;
+    return true;
+  static const unsigned UpperBoundOnSupplementaryOutputFileTypes =
+  types::TY_INVALID;
+  return InputActions.size() * UpperBoundOnSupplementaryOutputFileTypes >
```
…although I'm tempted to say we should just always do this when the number of primary inputs is greater than 1.

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -458,3 +461,75 @@ void SupplementaryOutputPathsComputer::deriveModulePathParameters(
+      {types::TY_SerializedDiagnostics, paths.SerializedDiagnosticsPath},
+      {types::TY_ModuleTrace, paths.LoadedModuleTracePath},
+      {types::TY_TBD, paths.TBDPath}};
+  for (const std::pair<types::ID, std::string &> *typeAndString =
```
You can still use for-each loops on variables statically typed as arrays! You just can't use them on pointers.

davidungar replied:
Cool! Done.

jrose-apple commented on lib/Frontend/OutputFileMap.cpp:
```diff
@@ -108,21 +102,47 @@ void OutputFileMap::dump(llvm::raw_ostream &os, bool Sort) const {
+    const TypeToPathMap *outputMap = getOutputMapForInput(input);
+    if (!outputMap)
+      continue;
+    os.write_escaped(input);
```
You still need quotes around the calls to `write_escaped`, sorry.

davidungar replied:
Thanks.

jrose-apple replied:
You *still* need quotes around the call to `write_escaped`. ;-) (You got the one below but this one was missed.)

jrose-apple commented on test/Driver/filelists.swift:
```diff
@@ -5,15 +5,28 @@
 // CHECK: Handled a.swift
+// CHECK-NEXT: Supplementary a.swift:
+// CHECK-NEXT: Supplementary a.swiftdoc
+// CHECK-NEXT: Supplementary a.swiftmodule
```
Shouldn't this be something like "Supplementary swiftdoc: a.swiftdoc"?

davidungar replied:
Good point! I'll write more Python.

jrose-apple reviewed (commented):
More comments. I'm thinking it would be a good idea to separate out the Frontend parts and commit those separately, since they're non-controversial. (They're also the parts that still need tests, though.)

davidungar commented on include/swift/Driver/Compilation.h:
```diff
@@ -167,6 +168,11 @@ class Compilation {
   unsigned BatchSeed;
 
+  /// In order to test repartitioning, set to true if
+  /// -force-one-batch-repartition is present. This is cleared after the forced
```
Thanks! Good catch.

davidungar commented on include/swift/Driver/Compilation.h:
```diff
@@ -272,6 +279,12 @@ class Compilation {
     return EnableBatchMode;
   }
 
+  bool getAndClearForceOneBatchRepartition() {
```
Ah... I see, we want one level of splitting each time, not one level of splitting once.
I'll look at changing this.

davidungar commented on include/swift/Frontend/OutputFileMap.h:
```diff
@@ -38,22 +39,20 @@ class OutputFileMap {
+    /// of the buffer in the process.
+    ///
+    /// When non-empty, \p workingDirectory is used to resolve relative paths in
+    /// the output file map.
```
Thanks--sorry!

davidungar commented on include/swift/Frontend/OutputFileMap.h:
```diff
@@ -38,22 +39,20 @@ class OutputFileMap {
-  loadFromPath(StringRef Path, StringRef workingDirectory);
+  /// Loads an OutputFileMap from the given \p Path into the receiver, if
+  /// possible.
+  static llvm::Expected<OutputFileMap> loadFromPath(StringRef Path,
```
No sweat! Doug pointed me to right place. (Thanks, @DougGregor .)

davidungar commented on include/swift/Frontend/OutputFileMap.h:
```diff
@@ -74,19 +73,19 @@ class OutputFileMap {
-  /// When non-empty, \p workingDirectory is used to resolve relative paths in
-  /// the output file map.
+  /// \brief Parses the given \p Buffer and returns either an OutputFileMap or
+  /// error string, taking ownership of \p Buffer in the process.
```
I *love* those nitpicks! Fixed.

davidungar commented on include/swift/Frontend/OutputFileMap.h:
```diff
@@ -74,19 +73,19 @@ class OutputFileMap {
-  bool parse(std::unique_ptr<llvm::MemoryBuffer> Buffer,
-             StringRef workingDirectory);
+  /// FIXME: Make the returned error strings more specific by including some of
+  /// the source.
```
Yes! Moved.

Commit:
Fixes

Commit:
JobContext name fixes

Commit:
Fix index.

Commit:
Fix SourceKit/CodeComplete/complete_playground_symlink.swift test.

Don't filter inputs by type for SingleCompile.

davidungar commented:
@swift-ci please smoke test

Commit:
Typo

davidungar commented:
@swift-ci please smoke test

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -52,6 +52,12 @@ bool ToolChain::JobContext::shouldUsePrimaryInputFileList() const {
+bool ToolChain::JobContext::shouldFilterInputsByType() const {
+  // SingleCompile's must not filter inputs by type in order for the
+  // playground (test) to work.
+  return OI.CompilerMode != OutputInfo::Mode::SingleCompile;
```
If you're going to put in a hack for playgrounds, please make it dependent on the `-playground` flag, not WMO-or-not. It'd be better if we could change the other side, of course.

davidungar replied:
This hack merely restores what the code did before. See calls to `types::isPartOfSwiftCompilation` in the relevant `constructInvocation` method. Is the comment misleading? What would be better?

Commit:
Filelist fixes.

Commit:
driver-compile.swift typo and format fix

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test

Commit:
Fix check-filelist-abc.py indent.

Commit:
Redo force repartion flag and add a test.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@jrose-apple @graydon OK, at this point, I *think* I've fixed all the test failures and addressed the prior suggestions, so it's likely ready for another pass. (I'll save any visitor-pattern-oidal work for the future.)

jrose-apple commented on lib/Frontend/ArgsToFrontendOutputsConverter.cpp:
```diff
@@ -458,3 +461,72 @@ void SupplementaryOutputPathsComputer::deriveModulePathParameters(
+    Diags.diagnose(
+        SourceLoc(),
+        diag::
+            error_cannot_have_supplementary_outputs_with_supplementary_ouputs_file_map,
```
Nitpick: I think you might as well shorten the name at this point. It's not buying us anything to repeat "supplementary outputs", for example. (Particularly with a typo in it ;-) )

davidungar replied:
Shortened. And I continue to be impressed by your ability to spot those details!

jrose-apple replied:
Well, it took me three times around to see the typo!

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -37,8 +38,52 @@ using namespace swift;
+
+bool ToolChain::JobContext::shouldFilterInputsByType() const {
+  // SingleCompile's must not filter inputs by type in order for the
+  // playground (test) to work.
```
Still don't think this is a good idea, because it's not the real condition.

davidungar replied:
If I remove this test, then WMO compiles will behave differently than before. This test will fail: SourceKit_CodeComplete/complete_playground_symlink_swift, but no other smoke tests, I believe. I agree that it's disturbing to break symmetry by not screening inputs for WMO compiles.  Do you think I should go to whomever would know about how playgrounds use the compiler and ask? Or what would you like me to do? (Maybe I'm misunderstanding your thoughts here.)

jrose-apple replied:
(We went and talked to Ben Langmuir about this and discovered that the current behavior is…well, sort of relied upon. I asked David to change the comment to say something more like 'SingleCompile used to behave like this, and some clients are relying on this (i.e. they pass inputs that don't have ".swift" as an extension)'.)

davidungar replied:
I've added a comment in a pidgin of Dave-ese and Jordan-ese. Can change it if you like.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -37,8 +38,52 @@ using namespace swift;
+  return Output.getPrimaryOutputFilenames().size() > TOO_MANY_FILES;
+}
+
+bool ToolChain::JobContext::shouldUseSupplementaryOutputFileList() const {
```
Should this say "map"? I know the enum is still called FilelistKind or whatever, but in this case we know what we're talking about.

jrose-apple replied:
This one should probably also have "frontend" in it for consistency.

davidungar replied:
Yes, indeed! Fixed: `shouldUseSupplementaryOutputFileMapInCompilerInvocation`. Tx.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -37,8 +38,52 @@ using namespace swift;
+  return Inputs.size() > TOO_MANY_FILES;
+}
+
+bool ToolChain::JobContext::shouldUseMainOutputFileList() const {
```
This should have the word "frontend" in it somewhere.

davidungar replied:
Fixed: `shouldUseMainOutputFileListInCompilerInvocation` as are the others.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+    ArgStringList &arguments) const {
+  llvm::StringSet<> primaries;
+
+  if (mayHavePrimaryInputs)
```
Nitpick: please put braces around multi-*line* if-statements, even if it's only a single *statement* in the body.

davidungar replied:
Fixed. (I should know better:)

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
+    ArgStringList &arguments) const {
+  // FIXME: get these and other argument strings from the same place for both
+  // driver and frontend
```
👍

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
+    ArgStringList &arguments) const {
+  // FIXME: get these and other argument strings from the same place for both
+  // driver and frontend
```
👍

jrose-apple replied:
(Oh, but please follow proper comment style with the initial capital letter and the trailing period.)

davidungar replied:
Thanks, great reminder!

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+  addSerializeDiagnosticsPathArgIfNecessary(Output, Args, arguments);
+
+  if (addOutputsOfType(arguments, Output, Args, types::ID::TY_ObjCHeader,
+                       "-emit-objc-header-path")) {
```
Why did you factor out `addModuleDocPathArgIfNecessary` but not `-emit-objc-header-path`?

davidungar replied:
`-emit-objc-header-path` is only created in `ToolChain::constructInvocation(const MergeModuleJobAction &job, const JobContext &context)`.

davidungar replied:
Thanks, great reminder!

jrose-apple replied:
That can't be right. It's needed for SingleCompile too.

jrose-apple replied:
Sorry, I mean "it's created here, for SingleCompile, and also in MergeModuleJobAction".

davidungar replied:
I am so glad you pointed this out! When I looked at it, I decided to make two changes:
1. Eliminate a level of (over) factoring. I decided that `addModuleDocPathArgIfNecessary` and its brethren were over-egging the pudding.
2. Add a bit of symmetry: compile jobs were emitting the objc header path via `addOutputsOfType` but the merge-module path was not. I changed the latter path to be symmetric with the former.

I'm hoping this arrangement will feel better.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+                   "-emit-reference-dependencies-path");
+  addOutputsOfType(arguments, Output, Args, types::TY_ModuleTrace,
+                   "-emit-loaded-module-trace-path");
+  addOutputsOfType(arguments, Output, Args, types::TY_TBD, "-emit-tbd-path");
```
…or `-emit-tbd-path`?

davidungar replied:
The TBDPath appears to be determined by `chooseTBDPath` which is only called for a `CompileJobAction`: 
      if (isa<CompileJobAction>(JA))
        chooseDependenciesOutputPaths(C, OI, OutputMap, workingDirectory, Buf,
                                  Output.get());
So it seemed unnecessary.

jrose-apple replied:
This one is needed for SingleCompile but currently not supported in StandardCompile, so I guess it's okay.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1705,12 +1723,13 @@ Driver::buildOutputFileMap(const llvm::opt::DerivedArgList &Args,
+                   llvm::toString(std::move(Err)));
   }
-  return OFM;
+  return llvm::make_unique<OutputFileMap>(*OFM);
```
Uh, hm. I guess if this is properly moving the OFM it's okay (i.e. we won't copy the whole hash map). Can you use `std::move(*OFM)` here?

jrose-apple replied:
(This is probably one of those things that doesn't matter in practice, but it is something I noticed.)

davidungar replied:
I've changed it to return an Optional<OutputFileMap>, but I'm not sure where I need the move's and where the compiler will do the optimization for me (partly because there's an implicit conversion to Optional).

jrose-apple replied:
The compiler will only do the optimization if it's the direct "argument" of `return`; pretty much never else. (I think there's a warning now for if you *do* use `std::move` when it's unnecessary, so you can err on the side of moving when there's a type big enough to warrant it.)

jrose-apple commented on include/swift/Frontend/OutputFileMap.h:
```diff
@@ -74,19 +73,16 @@ class OutputFileMap {
-  bool parse(std::unique_ptr<llvm::MemoryBuffer> Buffer,
-             StringRef workingDirectory);
+  /// \brief Parses the given \p Buffer and returns either an OutputFileMap or
+  /// error, taking ownership of \p Buffer in the process.
```
Huh, it looks like nothing actually needs this to take ownership of the buffer. Maybe that can go back to being a reference (later).

davidungar replied:
By "later", I'm guessing you mean "after this PR".

jrose-apple commented on include/swift/Driver/ToolChain.h:
```diff
@@ -83,6 +86,45 @@ class ToolChain {
     const char *getTemporaryFilePath(const llvm::Twine &name,
                                      StringRef suffix = "") const;
+
+    /// Test the number of files to see if an input file list is needed.
```
Thinking about it, I think these comments are at the wrong level:
- They should include what Action they're intended to be used for.
- They should *not* include that they're based on the number of files.
- They should not just echo the name of the function, which some of the ones below are doing. Better to have no comment in that case.

davidungar replied:
This is a great catch! Have taken a shot at improvement.

jrose-apple reviewed (commented):
More comments, but it is converging.

davidungar commented:
Thanks for looking at this again! I'll go through these, make fixes, and get back to you with any questions.

> On Mar 14, 2018, at 10:57 AM, Jordan Rose <notifications@github.com> wrote:
> 
> @jrose-apple commented on this pull request.
> 
> More comments, but it is converging.
> 
> In lib/Frontend/ArgsToFrontendOutputsConverter.cpp <https://github.com/apple/swift/pull/15013#discussion_r174541031>:
> 
> > +}
> +
> +Optional<std::vector<SupplementaryOutputPaths>>
> +SupplementaryOutputPathsComputer::readSupplementaryOutputFileMap() const {
> +  if (Arg *A = Args.getLastArg(options::OPT_emit_objc_header_path,
> +                               options::OPT_emit_module_path,
> +                               options::OPT_emit_module_doc_path,
> +                               options::OPT_emit_dependencies_path,
> +                               options::OPT_emit_reference_dependencies_path,
> +                               options::OPT_serialize_diagnostics_path,
> +                               options::OPT_emit_loaded_module_trace_path,
> +                               options::OPT_emit_tbd_path)) {
> +    Diags.diagnose(
> +        SourceLoc(),
> +        diag::
> +            error_cannot_have_supplementary_outputs_with_supplementary_ouputs_file_map,
> Nitpick: I think you might as well shorten the name at this point. It's not buying us anything to repeat "supplementary outputs", for example. (Particularly with a typo in it ;-) )
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174541320>:
> 
> > +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return getTopLevelInputFiles().size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUsePrimaryInputFileList() const {
> +  // SingleCompile's must not return true because then all inputs erroneously
> +  // end up in primary file list.
> +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return InputActions.size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldFilterInputsByType() const {
> +  // SingleCompile's must not filter inputs by type in order for the
> +  // playground (test) to work.
> Still don't think this is a good idea, because it's not the real condition.
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174541610>:
> 
> > +  return Inputs.size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseLinkInputFileList() const {
> +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return Inputs.size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseMainOutputFileList() const {
> +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return Output.getPrimaryOutputFilenames().size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseSupplementaryOutputFileList() const {
> Should this say "map"? I know the enum is still called FilelistKind or whatever, but in this case we know what we're talking about.
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174541707>:
> 
> > +  return OI.CompilerMode != OutputInfo::Mode::SingleCompile;
> +}
> +
> +bool ToolChain::JobContext::shouldUseMergeModuleInputFileList() const {
> +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return Inputs.size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseLinkInputFileList() const {
> +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return Inputs.size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseMainOutputFileList() const {
> This should have the word "frontend" in it somewhere.
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174541804>:
> 
> > +  return Inputs.size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseLinkInputFileList() const {
> +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return Inputs.size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseMainOutputFileList() const {
> +  if (Args.hasArg(options::OPT_driver_use_filelists))
> +    return true;
> +  return Output.getPrimaryOutputFilenames().size() > TOO_MANY_FILES;
> +}
> +
> +bool ToolChain::JobContext::shouldUseSupplementaryOutputFileList() const {
> This one should probably also have "frontend" in it for consistency.
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174542355>:
> 
> > @@ -526,6 +353,186 @@ ToolChain::constructInvocation(const CompileJobAction &job,
>    return II;
>  }
>  
> +const char *ToolChain::JobContext::computeFrontendModeForCompile() const {
> (We decided that it's okay to leave these on JobContext as long as they have "Frontend" in the name.)
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174546639>:
> 
> > +    Arguments.push_back("-supplementary-output-file-map");
> +    Arguments.push_back(getTemporaryFilePath("supplementaryOutputs", ""));
> +    FilelistInfos.push_back({Arguments.back(), types::TY_INVALID,
> +                             FilelistInfo::WhichFiles::SupplementaryOutput});
> +  } else {
> +    addFrontendSupplementaryOutputArguments(Arguments);
> +  }
> +}
> +
> +void ToolChain::JobContext::addFrontendInputArguments(
> +    const bool mayHavePrimaryInputs, const bool useFileList,
> +    const bool usePrimaryFileList, const bool filterByType,
> +    ArgStringList &arguments) const {
> +  llvm::StringSet<> primaries;
> +
> +  if (mayHavePrimaryInputs)
> Nitpick: please put braces around multi-line if-statements, even if it's only a single statement in the body.
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174547006>:
> 
> > +        arguments.push_back(inputName);
> +      }
> +    }
> +  }
> +}
> +
> +bool ToolChain::canCompileInputArgumentBePrimary(const CommandOutput &Output,
> +                                                 const llvm::opt::Arg &A) {
> +  return Output.getPrimaryOutputType() != types::TY_IndexData ||
> +         A.getOption().getID() == options::OPT_index_file_path;
> +}
> +
> +void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
> +    ArgStringList &arguments) const {
> +  // FIXME: get these and other argument strings from the same place for both
> +  // driver and frontend
> 👍
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174547079>:
> 
> > +        arguments.push_back(inputName);
> +      }
> +    }
> +  }
> +}
> +
> +bool ToolChain::canCompileInputArgumentBePrimary(const CommandOutput &Output,
> +                                                 const llvm::opt::Arg &A) {
> +  return Output.getPrimaryOutputType() != types::TY_IndexData ||
> +         A.getOption().getID() == options::OPT_index_file_path;
> +}
> +
> +void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
> +    ArgStringList &arguments) const {
> +  // FIXME: get these and other argument strings from the same place for both
> +  // driver and frontend
> 👍
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174547223>:
> 
> > +        arguments.push_back(inputName);
> +      }
> +    }
> +  }
> +}
> +
> +bool ToolChain::canCompileInputArgumentBePrimary(const CommandOutput &Output,
> +                                                 const llvm::opt::Arg &A) {
> +  return Output.getPrimaryOutputType() != types::TY_IndexData ||
> +         A.getOption().getID() == options::OPT_index_file_path;
> +}
> +
> +void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
> +    ArgStringList &arguments) const {
> +  // FIXME: get these and other argument strings from the same place for both
> +  // driver and frontend
> (Oh, but please follow proper comment style with the initial capital letter and the trailing period.)
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174547539>:
> 
> > +         A.getOption().getID() == options::OPT_index_file_path;
> +}
> +
> +void ToolChain::JobContext::addFrontendSupplementaryOutputArguments(
> +    ArgStringList &arguments) const {
> +  // FIXME: get these and other argument strings from the same place for both
> +  // driver and frontend
> +  addOutputsOfType(arguments, Output, Args, types::ID::TY_SwiftModuleFile,
> +                   "-emit-module-path");
> +
> +  addModuleDocPathArgIfNecessary(Output, Args, arguments);
> +
> +  addSerializeDiagnosticsPathArgIfNecessary(Output, Args, arguments);
> +
> +  if (addOutputsOfType(arguments, Output, Args, types::ID::TY_ObjCHeader,
> +                       "-emit-objc-header-path")) {
> Why did you factor out addModuleDocPathArgIfNecessary but not -emit-objc-header-path?
> 
> In lib/Driver/ToolChains.cpp <https://github.com/apple/swift/pull/15013#discussion_r174547618>:
> 
> > +  addSerializeDiagnosticsPathArgIfNecessary(Output, Args, arguments);
> +
> +  if (addOutputsOfType(arguments, Output, Args, types::ID::TY_ObjCHeader,
> +                       "-emit-objc-header-path")) {
> +    assert(OI.CompilerMode == OutputInfo::Mode::SingleCompile &&
> +           "The Swift tool should only emit an Obj-C header in single compile"
> +           "mode!");
> +  }
> +
> +  addOutputsOfType(arguments, Output, Args, types::TY_Dependencies,
> +                   "-emit-dependencies-path");
> +  addOutputsOfType(arguments, Output, Args, types::TY_SwiftDeps,
> +                   "-emit-reference-dependencies-path");
> +  addOutputsOfType(arguments, Output, Args, types::TY_ModuleTrace,
> +                   "-emit-loaded-module-trace-path");
> +  addOutputsOfType(arguments, Output, Args, types::TY_TBD, "-emit-tbd-path");
> …or -emit-tbd-path?
> 
> In lib/Driver/Job.cpp <https://github.com/apple/swift/pull/15013#discussion_r174548379>:
> 
> > @@ -240,6 +240,10 @@ StringRef CommandOutput::getAnyOutputForType(types::ID Type) const {
>    return getAdditionalOutputForType(Type);
>  }
>  
> +const OutputFileMap &CommandOutput::getDerivedOutputMap() const {
> +  return DerivedOutputMap;
> Maybe the whole "write the subset of the DOM based on this Job's inputs" should be pushed into Job (taking an llvm::raw_ostream &). What do you think?
> 
> In lib/Driver/Driver.cpp <https://github.com/apple/swift/pull/15013#discussion_r174550043>:
> 
> >    }
> -  return OFM;
> +  return llvm::make_unique<OutputFileMap>(*OFM);
> Uh, hm. I guess if this is properly moving the OFM it's okay (i.e. we won't copy the whole hash map). Can you use std::move(*OFM) here?
> 
> In lib/Driver/Driver.cpp <https://github.com/apple/swift/pull/15013#discussion_r174550149>:
> 
> >    }
> -  return OFM;
> +  return llvm::make_unique<OutputFileMap>(*OFM);
> (This is probably one of those things that doesn't matter in practice, but it is something I noticed.)
> 
> In lib/Driver/Driver.cpp <https://github.com/apple/swift/pull/15013#discussion_r174550435>:
> 
> > @@ -1102,7 +1118,8 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
>  
>    } else { // DriverKind::Batch
>      OI.CompilerMode = OutputInfo::Mode::StandardCompile;
> -    if (Args.hasArg(options::OPT_whole_module_optimization))
> +    if (Args.hasArg(options::OPT_whole_module_optimization,
> +                    options::OPT_index_file))
> I really think the right answer is to have -index-file force single-file mode even if the user passes -wmo, and then -index-file-path controls the one job that gets made.
> 
> In include/swift/Frontend/OutputFileMap.h <https://github.com/apple/swift/pull/15013#discussion_r174552262>:
> 
> >  private:
> -  /// \brief Parses the given \p Buffer into the OutputFileMap, taking ownership
> -  /// of \p Buffer in the process.
> -  ///
> -  /// When non-empty, \p workingDirectory is used to resolve relative paths in
> -  /// the output file map.
> -  ///
> -  /// \returns true on error, false on success
> -  bool parse(std::unique_ptr<llvm::MemoryBuffer> Buffer,
> -             StringRef workingDirectory);
> +  /// \brief Parses the given \p Buffer and returns either an OutputFileMap or
> +  /// error, taking ownership of \p Buffer in the process.
> Huh, it looks like nothing actually needs this to take ownership of the buffer. Maybe that can go back to being a reference (later).
> 
> In include/swift/Driver/ToolChain.h <https://github.com/apple/swift/pull/15013#discussion_r174553736>:
> 
> > @@ -83,6 +86,45 @@ class ToolChain {
>      /// arguments.
>      const char *getTemporaryFilePath(const llvm::Twine &name,
>                                       StringRef suffix = "") const;
> +
> +    /// Test the number of files to see if an input file list is needed.
> Thinking about it, I think these comments are at the wrong level:
> 
> They should include what Action they're intended to be used for.
> They should not include that they're based on the number of files.
> They should not just echo the name of the function, which some of the ones below are doing. Better to have no comment in that case.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift/pull/15013#pullrequestreview-103917235>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAcJ0JKWVrWNc3TPnfxnToGUZLpBCrHIks5teVoIgaJpZM4SeKUm>.
>

Commit:
Fix comments and names

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -986,9 +1001,11 @@ void Driver::buildInputs(const ToolChain &TC,
+      if (checkInputExistence(*this, Args, Diags, Value)) {
+        // Substitute the OPT_index_file_path Arg if this is one of them.
+        const Arg *IA = IndexFilePathArgs.lookup(Value);
+        Inputs.push_back(std::make_pair(Ty, IA ? IA : A));
```
New round of comments where I've given up on improving `-index-file`: I think this is too clever. Yes, it gets the information where you want it to go, but at the cost of making `-index-file-path` "input-like", which it really wasn't before. I'd rather just go back to the list of arguments and see if the name matches up later.

davidungar replied:
I have pushed a version with index file path handling that I hope is better.

jrose-apple commented on lib/Driver/Compilation.cpp:
```diff
@@ -1122,16 +1142,29 @@ static bool writeFilelistIfNecessary(const Job *job, DiagnosticEngine &diags) {
+        const llvm::opt::Arg &InArg = IA->getInputArg();
+        if (ToolChain::canCompileInputArgumentBePrimary(job->getOutput(),
+                                                        InArg))
+          out << IA->getInputArg().getValue() << "\n";
```
New round of comments where I've given up on improving `-index-file`: …that is, here I'd just put a method on JobContext called `isFrontendPrimaryInput` or something that takes an InputAction and checks to see if Args has `OPT_index_file_path` and if it matches up.

jrose-apple replied:
(Because the *right* answer is that `-index-file` is neither a StandardCompile nor a BatchCompile nor a SingleCompile; it's a OneInputOnlyCompile or something.)

davidungar replied:
I looked a lot at making the index-file-path handling less clever, and it gets very ugly. The problem is that the functionality is needed in several places: When constructing a command-line, when building a filelist from both `performJobImpl`, and in `addPendingJobToTaskQueue`. So one ends up passing arguments around. Then the logic to select a primary argument gets complex enough that I tried to factor it out. Then that gets hairy. I can take an other run at it tomorrow, but the clever solution has the appeal of centralizing the hack, and simplifying the downstream.

As an aside, the status quo (and this PR) duplicates the logic used to enumerate inputs and outputs between file list construction and command line creation. It used to be just inputs and outputs, but now it's inputs, primary inputs, main outputs, and supplementary outputs.
It's crying for a fix: either factoring or moving to just using filelists, as we've spoken about.

Let me know what you think or if you want to talk more offline, thanks!

jrose-apple replied:
Sorry, why do we need it for `addPendingJobToTaskQueue`?

davidungar replied:
I suspect it was added by @graydon as part of the make-batch-jobs-work-for-incremental-compilation work. It's called by `transferJobsToTaskQueue`, which is called by `formBatchJobsAndAddPendingJobsToTaskQueue`, which is called in two places: the first is likely as before: `performJobsImpl`, but the second is more interesting: it's called from `runTaskQueueToCompletion` and the comment is:
        // Task-exit callbacks from TaskQueue::execute may have unblocked jobs,
        // which means there might be PendingExecution jobs to enqueue here. If
        // there are, we need to continue trying to make progress on the
        // TaskQueue before we start marking deferred jobs as skipped, below.
So, what I deduce is going on is that batch job formation happens *after* the scheduling for incremental jobs, and that's why things are called from two places.
The idea AFAICT is to lazily do the batching so that we can more optimally schedule the compilation of individual primaries in batches.

Commit:
Return an Optional OutputFileMap instead of a unique pointer.

davidungar commented:
Regarding your comment "Maybe the whole "write the subset of the DOM based on this Job's inputs" should be pushed into Job (taking an llvm::raw_ostream &). What do you think?": (I don't seem to have a reply button in the right place, so I'll reply here: 
If you're toying with the idea of moving `OutputFileMap::write`, I would tend to keep it on OutputFileMap.cpp, because it knows about the in-memory-to-external translation, as does the existing `OutputFileMap::parse` method. If you meant moving all or part of `writeFilelistIfNecessary` then I like the idea of moving that whole routine into `Job`. Except I tried it, and drags some stuff into Job (like diagnostics, and `canCompileInputArgumentBePrimary`, so I'll leave it for now.

Commit:
Small cosmetic fixes.

davidungar commented:
@swift-ci please smoke test

Commit:
Error typo fix.

Commit:
OutputFileMap type fix.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test

jrose-apple commented:
> If you're toying with the idea of moving `OutputFileMap::write`, I would tend to keep it on OutputFileMap.cpp, because it knows about the in-memory-to-external translation, as does the existing `OutputFileMap::parse` method. If you meant moving all or part of `writeFilelistIfNecessary` then I like the idea of moving that whole routine into Job. Except I tried it, and drags some stuff into Job (like diagnostics, and `canCompileInputArgumentBePrimary`, so I'll leave it for now.

I was thinking closer to the latter:

```c++
void Job::writeOutputFileMap(llvm::raw_ostream &out) const {
  // Compute input names here
  getDerivedOutputFileMap().write(out, names);
}
```

but I see how there's a bit of a problem with `canCompileInputArgumentBePrimary` again. Maybe that's just an extra argument though, `StringRef pathToForceAsPrimary`.

(Actually, if `-index-file` is only used in SingleCompile mode, then we know we only need one supplementary file map entry anyway today. And on *top* of that we know there's only one entry, so we don't really need the file map at all. Maybe the simple thing to do is say "if we're doing index file stuff, don't use a supplementary file map".)

Commit:
Remove overfactoring and restore symmetry for -emit-objc-header-path

davidungar commented:
@swift-ci please smoke test

davidungar commented:
Have implemented your `writeOutputFileMap` suggestion. Love it!

Commit:
Factor out writeOutputFileMap

davidungar commented:
@swift-ci please smoke test

Commit:
Typo fix for -emit-objc-header-path.

davidungar commented:
@swift-ci please smoke test

Commit:
Undo buildInputs index hack

jrose-apple commented on include/swift/Driver/Util.h:
```diff
@@ -46,7 +46,11 @@ namespace driver {
+      /// Used in batch mode to provide supplementary output file names.
+      /// An oddball because there are lots of paths and
+      /// types, uses the whole DerivedOutputFileMap.
+      SupplementaryOutput,
```
David and I have gone back and forth about this. It would mean switching WMO over to using the output file map as well, which is a little weird because WMO only has one of each supplementary output.

davidungar replied:
I would like to subsume *all* filename arguments to the frontend in one file map, and use helpers for testing. I think the frontend code might be more maintainable with only one way to do things. Jordan & I have talked about this, but not very much. Not the fish to fry right now.

Commit:
Fix supplementary output filemap generation to account for CommandOutput.

Commit:
Rename addFrontendInputArguments.

Commit:
Handle index file path differently.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@graydon I think I've incorporated the fix to writing OutputFileMap that you pointed out today.

davidungar commented:
@jrose-apple I have redone the handling of -index-file-path.

davidungar commented on lib/Driver/Compilation.cpp:
```diff
@@ -1122,16 +1142,29 @@ static bool writeFilelistIfNecessary(const Job *job, DiagnosticEngine &diags) {
+        const auto *IA = cast<InputAction>(A);
+        inputs.push_back(IA->getInputArg().getValue());
+      }
+      job->getOutput().getDerivedOutputMap().write(out, inputs);
```
Fixed as per our conversation. Thanks a lot for spotting it.

davidungar commented on lib/Driver/ToolChains.cpp:
```diff
@@ -342,6 +350,14 @@ ToolChain::constructInvocation(const CompileJobAction &job,
                                   FilelistInfo::WhichFiles::PrimaryInputs});
     }
+    if (UseSupplementaryOutputFileList) {
+      Arguments.push_back("-supplementary-output-filelist");
```
It's now called supplementary-output-file-map, and only used for supplementary outputs.
After this PR, we might want to unify things.

davidungar commented:
@jrose-apple At present, I believe that -index-file is forcing single-compile mode. I can try having it force single-file mode instead if you like. But that may tangle things up a bit when we fold single-file into batch mode.

jrose-apple commented:
I like this handling of -index-file a lot better. Thank you! (And yeah, I've given up on making it not inject a primary file right at the end.)

davidungar commented on lib/Driver/ToolChains.cpp:
```diff
@@ -37,8 +38,41 @@ using namespace swift;
+  static const unsigned UpperBoundOnSupplementaryOutputFileTypes =
+      types::TY_INVALID;
+  return InputActions.size() * UpperBoundOnSupplementaryOutputFileTypes >
+  TOO_MANY_FILES;
```
Fixed.

davidungar commented on lib/Driver/Driver.cpp:
```diff
@@ -2650,3 +2656,19 @@ void Driver::printHelp(bool ShowHidden) const {
+bool OutputInfo::mightHaveExplicitPrimaryInputs(
+    const CommandOutput &Output) const {
+  // SingleCompiles (WMO's) must not return true because then all inputs
+  // erroneously end up in primary file list.
```
Deleted.

davidungar commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +374,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+      arguments.push_back("-primary-file");
+      arguments.push_back(inputName);
+    }
+    if (!isPrimary && !useFileList)
```
Thanks. Fixed.

davidungar commented on lib/Driver/ToolChains.cpp:
```diff
@@ -526,6 +374,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+                             FilelistInfo::WhichFiles::PrimaryInputs});
+  }
+  if (!UseFileList || !UsePrimaryFileList) {
+    addFrontendCommandLineInputArguments(MayHavePrimaryInputs, UseFileList,
```
Yes, I could. I like it as is because it's a bit clearer this way.

davidungar commented on lib/Driver/ToolChains.cpp:
```diff
@@ -37,8 +38,41 @@ using namespace swift;
+
+bool ToolChain::JobContext::shouldFilterCompilerInputsByType() const {
+  // SingleCompile's must not filter inputs by type in order for the
+  // playground (test) to work.
```
Changed with your help, tx.

davidungar commented on lib/Driver/Job.cpp:
```diff
@@ -310,6 +314,16 @@ CommandOutput::dump() const {
 }
 
+void CommandOutput::writeOutputFileMap(llvm::raw_ostream &out) const {
+  std::vector<StringRef> inputs;
```
Done!

davidungar commented on include/swift/Driver/ToolChain.h:
```diff
@@ -83,6 +86,35 @@ class ToolChain {
+  private:
+    void addFrontendCommandLineInputArguments(
+        const bool mayHavePrimaryInputs, const bool useFileList,
+        const bool usePrimaryFileList, const bool filterByType,
```
Oops! Fixed.

davidungar commented on include/swift/Driver/ToolChain.h:
```diff
@@ -83,6 +86,35 @@ class ToolChain {
+
+    bool shouldUsePrimaryInputFileListInFrontendInvocation() const;
+
+    bool shouldUseMainOutputFileListInCompilerInvocation() const;
```
Changed to "Frontend".

Commit:
Review fixes, some cosmetic, some clarifications, one for an improbable situation.

davidungar commented:
@swift-ci please smoke test

jrose-apple commented on include/swift/Driver/ToolChain.h:
```diff
@@ -83,6 +86,34 @@ class ToolChain {
+    /// Reify the existing behavior that SingleCompile compile actions do not
+    /// filter, but batch-mode and single-file compilations do. Some clients are
+    /// relying on this (i.e., they pass inputs that don't have ".swift" as an
+    /// extension.) It would be nice to eliminate this distinction someday.
```
Aha, this is probably the comment we were talking about! This form is fine, although it looks like an implementation comment rather than a declaration comment.

davidungar commented:
@swift-ci please smoke test

graydon reviewed (approved):
Thanks

Commit:
Fix test to reflect quotes on inputs in supplementary file map.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please test source compatibility
